File: csharp-playwright-nunit-reqnroll/SauceDemo.UiTests/TestData/InvalidCredentialsFactory.cs
using Bogus;
using SauceDemo.UiTests.Users;

namespace SauceDemo.UiTests.TestData;

/// <summary>Generates random credentials for negative login scenarios, guaranteed not to collide with a real UserCatalog account.</summary>
public static class InvalidCredentialsFactory
{
    public static (string Username, string Password) Create()
    {
        var faker = new Faker();

        string username;
        do
        {
            username = faker.Internet.UserName();
        } while (UserCatalog.AllUsers.ContainsKey(username));

        var password = faker.Internet.Password(length: 12);
        return (username, password);
    }
}

File: java-selenium-junit5-cucumber/src/test/java/com/saucedemo/uitests/testdata/InvalidCredentialsFactory.java
package com.saucedemo.uitests.testdata;

import com.saucedemo.uitests.users.UserCatalog;
import net.datafaker.Faker;

/**
 * Generates random credentials for negative login scenarios, guaranteed not to collide with a real
 * {@link UserCatalog} account.
 */
public final class InvalidCredentialsFactory {

    private static final int PASSWORD_LENGTH = 12;

    private InvalidCredentialsFactory() {
    }

    public record Credentials(String username, String password) {
    }

    public static Credentials create() {
        Faker faker = new Faker();

        String username = faker.internet().username();
        while (UserCatalog.ALL_USERS.containsKey(username)) {
            username = faker.internet().username();
        }

        return new Credentials(username, faker.internet().password(PASSWORD_LENGTH, PASSWORD_LENGTH + 1));
    }
}
